### PR TITLE
{Keyvault} Fix #22540: Create keyvault with service principal

### LINKED ISSUE #22540: Could not create keyvault with service principal after graph module migration
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Related command**

`az keyvault create`

**Describe the bug**

Could not create keyvault with service principal after graph module migration.

image

**To Reproduce**

Create keyvault with a service principal.

**Expected behavior**

Able to create keyvault with a service principal like it used to be.

**Environment summary**

WSL2, Ubuntu18.04, Python3.8, CLI `dev` branch + `tags/azure-cli-2.36.0`

**Additional context**

Before the migration, function [_get_current_user_object_id](https://github.com/Azure/azure-cli/blob/azure-cli-2.36.0/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L753) would raise GraphErrorException and retry with [_get_object_id](https://github.com/Azure/azure-cli/blob/azure-cli-2.36.0/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L755).

Screenshot 2022-05-20 152110

After the migration, function [_get_current_user_object_id](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L748) would not raise an error. Without retry, empty object id would result in raising [CLIError](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L752).

Screenshot 2022-05-20 152633

### PR DESCRIPTION
**Pull request by jiasli**
**Related command**
`az keyvault create`

**Description**

- Fix #22540

https://github.com/Azure/azure-cli/pull/22203 changed `msrestazure.azure_exceptions.CloudError` to `azure.cli.command_modules.role.GraphError` in `azure.cli.command_modules.keyvault.custom._get_current_user_object_id`.

Some background: **`azure-graphrbac` Python SDK raises inconsistent exceptions for Graph errors**:

`DomainsOperations.get` raises `CloudError`:

https://github.com/Azure/azure-sdk-for-python/blob/dfa109c36598d93aaf74238feded516352c20f4c/sdk/graphrbac/azure-graphrbac/azure/graphrbac/operations/domains_operations.py#L150-L153

```py
        if response.status_code not in [200]:
            exp = CloudError(response)
            exp.request_id = response.headers.get('x-ms-request-id')
            raise exp
```

`SignedInUserOperations.get` raises `GraphErrorException`:

https://github.com/Azure/azure-sdk-for-python/blob/dfa109c36598d93aaf74238feded516352c20f4c/sdk/graphrbac/azure-graphrbac/azure/graphrbac/operations/signed_in_user_operations.py#L79-L80

```py
        if response.status_code not in [200]:
            raise models.GraphErrorException(self._deserialize, response)
```

Since `SignedInUserOperations.get` doesn't raise `CloudError`, this `except CloudError` will never be hit:

https://github.com/Azure/azure-cli/blob/1451bd570fd0fe54771cedaecdbd49cdb4b8d69c/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L312-L317

**Therefore, this `except CloudError` clause has no effect and should NOT be replaced by `GraphError`.** Doing so will make `GraphError` be intercepted and the outer `except GraphError` have no effect:

https://github.com/Azure/azure-cli/blob/cac3f06cf483d8c81f93c9cd556116e27aff2182/src/azure-cli/azure/cli/command_modules/keyvault/custom.py#L747-L750

This PR lets `_get_current_user_object_id` raise `GraphError` so that it can be caught by outer `except GraphError`.

**Testing Guide**
```sh
az ad sp create-for-rbac --role Owner --scopes /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590 --verbose

# Copy the login command from log and run:
az login --service-principal --username 558c5061-b77b-4dfa-8118-168068c9be4a --password xxx --tenant 54826b22-38d6-4fb2-bad9-b7b93a3e9c5a

az group create -n my-kv-testrg -l eastasia
az keyvault create -n my-kv-test -g my-kv-testrg
```

### CODE REVIEW

**Comment by evelyn-ys** [comment]
Fix #22540

**Review by evelyn-ys — APPROVED** [review]

**Review by calvinhzy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 1]
```diff
@@ -304,15 +304,9 @@ def list_vault(client, resource_group_name=None):
     return list(vault_list)
 
 
-# pylint: disable=inconsistent-return-statements
 def _get_current_user_object_id(graph_client):
-    from azure.cli.command_modules.role import GraphError
-    try:
-        current_user = graph_client.signed_in_user_get()
-        if current_user and current_user.get('id', None):  # pylint:disable=no-member
-            return current_user['id']  # pylint:disable=no-member
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:313]: I did leave a comment for this weird behavior of checking the existence of `id` (https://github.com/Azure/azure-cli/pull/22203#discussion_r859747654). 

`graph_client.signed_in_user_get` will ALWAYS return a user object which MUST have `id` property UNLESS `GraphError` is raised.

Takeaways:

1. Pylint error is a sign of abnormality and should NOT be disabled brutally.
2. For anything unusual like `current_user.get('id', None)` or `except CloudError`, comments MUST be left to explain its behavior.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/keyvault/custom.py:313]: We got tripped by some code written almost 6 years ago (cb0faa82bc5db56a15d0c03e4dad5cf040ae9d83)!

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 2]
```diff
@@ -304,15 +304,9 @@ def list_vault(client, resource_group_name=None):
...
-    except GraphError:
-        pass
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:315]: We got tripped by some code written almost 6 years ago (cb0faa82bc5db56a15d0c03e4dad5cf040ae9d83)!

**Comment by yonzhan** [comment]
Good catch !

**Comment by jiasli** [comment]
I have confirmed the new RPM has the correct code:

```
> docker run -it --rm mcr.microsoft.com/cbl-mariner/base/core:2.0 bash

# tdnf install https://azurecliprod.blob.core.windows.net/archive/20220520.4/rpm-mariner2.0/signed/azure-cli-2.37.0-1.x86_64.rpm
# tdnf install sed
# cat /lib64/az/lib/python3.9/site-packages/azure/cli/command_modules/keyvault/custom.py | sed -n '307,309p'
def _get_current_user_object_id(graph_client):
    current_user = graph_client.signed_in_user_get()
    return current_user['id']
```

**Pull request merged**